kong/cmd/migrations.lua

```diff
@@ -94,21 +95,78 @@ local function execute(args)
 
   if args.command == "list" then
     if schema_state.needs_bootstrap then
-      log("database needs bootstrapping; run 'kong migrations bootstrap'")
-      os.exit(3)
+      if schema_state.legacy_invalid_state then
+        -- legacy: migration from 0.14 to 1.0 cannot be performed
+        if schema_state.legacy_missing_component then
+          log("Migrations can only be listed on a %s %s that has been " ..
+              "upgraded to 1.0, but the current %s seems to be older "  ..
+              "than 0.14 (missing migrations for '%s').",
+              db.strategy, db.infos.db_desc, db.infos.db_desc,
+              schema_state.legacy_missing_component)
+
+          os.exit(2)
+        end
+
+        if schema_state.legacy_missing_migration then
+          log("Migrations can only be listed on a %s %s that has been " ..
+              "upgraded to 1.0, but the current %s seems to be older "  ..
+              "than 0.14 (missing migration '%s' for '%s').",
+              db.strategy, db.infos.db_desc, db.infos.db_desc,
+              schema_state.legacy_missing_migration,
+              schema_state.legacy_missing_component)
+
+          os.exit(2)
+        end
+
+        log("Migrations can only be listed on a %s %s that has been "     ..
+            "upgraded to 1.0, but the current %s seems to be older than " ..
+            "0.14 (missing migrations).", db.strategy, db.infos.db_desc,
+            db.infos.db_desc)
+
+        os.exit(2)
+
+      elseif not schema_state.legacy_is_014 then
+        log("database needs bootstrapping; run 'kong migrations bootstrap'")
+        os.exit(3)
+      end
     end
 
+    local r = ""
+
     if schema_state.executed_migrations then
       log("executed migrations:\n%s", schema_state.executed_migrations)
+      r = "\n"
+    end
+
+    if schema_state.pending_migrations then
+      log("%spending migrations:\n%s", r, schema_state.pending_migrations)
+      r = "\n"
+    end
+
+    if schema_state.new_migrations then
+      log("%snew migrations available:\n%s", r, schema_state.new_migrations)
+      r = "\n"
     end
 
-    migrations_utils.print_state(schema_state)
+    if schema_state.pending_migrations and schema_state.new_migrations then
+      log("%s[warn] database has pending migrations from a previous upgrade, " ..
```

Comment:
There exists `log.warn` to set the proper logging level. In verbose mode, this would show up as `[notice] [warn]` (and logging level would not be respected for this warning).

Reply:
I know, but it puts timestamp (I can revert to that though).

Reply:
Yes, better use `log.warn` here (the newline can be put after the last line of the previous call to `log()`). Another option is to subsequently patch the logging utility to output a non-verbose log on `warn` and above levels (it currently is the desired behavior).